KEY-84 - Update BsonADT to support Mongo3

jeffmay opened the pull request:
https://jira.audaxhealth.com/browse/KEY-84

@egprentice I've created this bson-adt-mongo library so that we can use the latest Java driver with a nicer Scala syntax. Also, there are some advances we can make with this ADT. Basically, we could potentially copy and paste the play-json macros and use them for `BsonValue` types to get macro generated Mongo serialization. Either way, I didn't want to have to migrate later, so I just did this one first.

@audax-shreyaspurohit You might like Mongo's plain Java driver. Not that you aren't already committed to Reactive Mongo, but I actually like the v3 client interface a lot. You can swap out the underlying serialization mechanism (as I've done in this PR) to whatever type you like. Of course, I'm using my bson-adt :smile: 

@rwhiten1 Since you are interested in typeclasses. Here is an instance where I am not implementing the full spectrum of Mongo 3 client types (see http://docs.mongodb.org/manual/reference/bson-types/ for the full list of types). I still have type-safety since there is no way to create a `BsonValue` that has no mapping to `JavaBsonValue`, all `BsonValue` types are matched on. However, not all `JavaBsonValue`s have a mapping from `BsonValue`. This means if you go from `JavaBsonValue` to `BsonValue`, but are not going from `BsonValue` to `JavaBsonValue`, you could get an error. I plan to correct for that in future versions by supporting more types. I do support symmetrical serialization (as you can see in the property specs)

Commit:
Moved code into correct projects. Minor cleanup.

- Split bson-adt-core into bson-adt-mongo2
- Moved MongoDB 2 stuff from adt.bson to adt.bson.mongodb
- Created type aliases for backward compatability in bson-adt-mongo2
- Removed unnecessary imports
- Added BsonReads for ADT types in DefaultBsonReads and DefaultBsonWrites
- Convert BsonValue to use BsonType.
- Fixed function inheritence for BsonDate.
- Fixed DBObject imports.

Commit:
Split project into multi-project build.

Commit:
Added mongo 3.0.X codecs and generators

- Split up generators for better inheritence
- Added BsonAdtCodec and tests
- Added BsonAdtCodecProvider and tests for MongoDB
- Added implicit operations
- Updated some comments

jeffmay commented on bson-adt-core/src/test/scala/adt/bson/scalacheck/CommonGenerators.scala:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  // The only way in which the ObjectId could cause a property to fail is if someone is using
+  // the timestamp from the ObjectId in a test.
```
This comment could be made clearer.

jeffmay commented on bson-adt-core/src/test/scala/adt/bson/scalacheck/BsonValueGenerators.scala:
```diff
@@ -167,11 +113,12 @@ trait BsonValueGenerators extends RegexGenerators {
+//    shrink[Map[String, BsonValue]](bson.value) map { fields =>
+//      BsonObject(fields)
+//    }
+//  }
```
This was sometimes causing an infinite loop in the tests. I'll see if this is still the case.

jeffmay commented on build.sbt:
```diff
@@ -1,47 +1,87 @@
+  )
+).dependsOn(
+    `bson-adt-core` % "compile->compile;test->test"
+  )
```
Git is all confused because I made this a multi-project build.

Basically `bson-adt` split into:
1. `bson-adt-core` with only a dependence on `org.mongodb:bson:3.0.0`
2. `bson-adt-mongo2` with dependence on `bson-adt-core` and `org.mongodb:mongo-java-driver:2.13.1`
3. `bson-adt-casbah` with a dependence on `bson-adt-mongo2` and casbah
4. `bson-adt-mongo` with a dependence on `bson-adt-core` and `org.mongodb:mongo-java-driver:3.0.2`

I figure this is alright since `org.mongodb` split out the `bson` library in v3, so in all future versions, `bson-adt-mongo` would be the hub of any new extensions (include mongodb drivers themselves), but the bson version could remain the lowest necessary.

chrissalij-r replied:
:+1:

jeffmay commented:
Also, @danstaley @patrick-bohan @chrissalij-r @KyleMessner @jtsmith0107 because... well, I think you might be interested in this one.

Feel free to tag anyone else who might be interested in using the Mongo 3 clients. If we wanted to use this to smooth the migration path, people can switch to using `BsonFormat` for their mongo serialization logic, and it will just seemlessly work with Mongo 3 client's generic using interface. This library also supports a Casbah implementation for another project in production that runs Mongo 2.4 using the same ADT. We might consider using this library for serialization to allow cross-compatibility between Mongo driver versions (including Casbah and down the road, potentially, ReactiveMongo).

@bookest we can use this same library to drop down to Mongo 2.4 or 2.6 (it is used in another large project running in production this way). I would rather we test out Mongo 3 tho (thus why I did this implementation). But we should decide on whether going live with Mongo 3 is feasible. If we decide we need to drop to lib-mongo + casbah + mongo 2.6 (or 2.4), we don't have to rewrite the serialization logic.

egprentice commented on bson-adt-core/src/main/scala/adt/bson/BsonValue.scala:
```diff
@@ -1,8 +1,8 @@
+import org.bson.BsonType
 import org.bson.types.ObjectId
-import org.joda.time.DateTime
+import org.joda.time.{DateTime, DateTimeZone}
```
consider using nscala-time which is a nice scala wrapper around joda time
https://github.com/nscala-time/nscala-time

jeffmay replied:
I would consider it for a more intensive use of DateTime. Here, I am just throwing them into Bson and reading them out again. Nothing fancy.

egprentice commented on build.sbt:
```diff
@@ -1,47 +1,87 @@
+  libraryDependencies ++= Seq(
+    "joda-time" % "joda-time" % "2.8",
+    "org.joda" % "joda-convert" % "1.7",
+    "org.scalacheck" %% "scalacheck" % "1.11.3" % "test",
```
latest scalacheck is 1.12.4
https://www.scalacheck.org/download.html

jeffmay replied:
Cool. I'll upgrade on next release.

egprentice commented:
lgtm

KyleMessner commented on bson-adt-mongo2/src/main/scala/adt/bson/mongodb/DBValueExtractors.scala:
```diff
@@ -23,11 +24,11 @@ trait DBValueExtractors {
 
 trait DefaultDBValueExtractor {
-  x: DBValueExtractors =>
+  self: DBValueExtractors =>
```
what's wrong with just `x`? :trollface:

jeffmay replied:
It was shadowing other parameters... but, I think this is a little better

KyleMessner commented:
lgtm

chrissalij-r commented on bson-adt-casbah/src/main/scala/adt/bson/casbah/CasbahImplicits.scala:
```diff
@@ -24,7 +27,7 @@ trait CasbahImplicits {
   implicit def toDBObjectBsonOps(dbo: MongoDBObject): DBObjectBsonOps = new DBObjectBsonOps(dbo)
 
-  implicit def toDBObjectBsonOps(dbo: DBObject): DBObjectBsonOps = new DBObjectBsonOps(dbo)
+  implicit def toDBObjectBsonOps(dbo: com.mongodb.DBObject): DBObjectBsonOps = new DBObjectBsonOps(dbo)
```
Why are you specifying the full path?

jeffmay replied:
There is a `DBObject` extractor in the same package. It used to work fine because DBObject wasn't in the same package. When I moved this stuff into the Casbah package, it became a namespace conflict.

ghost commented:
LGTM

jeffmay commented on bson-adt-mongo3/src/test/scala/adt/bson/scalacheck/JavaBsonValueGenerators.scala:
```diff
@@ -0,0 +1,199 @@
+  // TODO: Fix this
+//  implicit val shrinkJavaBsonDocument: Shrink[JavaBsonDocument] = Shrink { bson =>
+//    shrink[Map[String, JavaBsonValue]](bson.values.toMap) map JavaBsonDocument
+//  }
```
I'll test this one too.

Commit:
Updates to BsonValue ADT

- Added BsonJavaScript and BsonTimestamp as placeholders.
- Updated Type val to def

audax-shreyaspurohit commented:
post merge lgtm
